### DISCUSSION TITLE
Which mode kube-proxy works on?

### DISCUSSION THREAD

**Discussion by matrix-root** [mysticaltech/terraform-hcloud-kube-hetzner#587] [Q&A] [answered]
Hi, dear hetzner-lovers! 😈

I've was very interested and tried to deep dive into k3s details and did not found any info about kube-proxy mode inside `/etc/rancher/k3s/config.yaml`

I've tried to check if iptables here - it's not installed. Tried to check ipvs - it's here and loaded as part of core, but if I'll try to install ipvsadm - I'll not see any rules, so looks like k3s kube-proxy does not use that

I don't know another way how to find mode in which k3s works :D 
Looks like it uses iptables (by default), but it's even not installed...

It's single-cluster setup with klipper, but it should be same at the multi-node as I undertsand

**Comment by mysticaltech** [comment, marked as answer]
@matrix-root k3s will use its own iptables utility, which ships along with its own binary. 

Some info (from ChatGPT):

#### You:
How does kube-proxy work inside of k3s

#### ChatGPT:
Kube-proxy is a component of Kubernetes that is responsible for managing network traffic between services within a cluster. In K3s, which is a lightweight distribution of Kubernetes, kube-proxy works in a similar way as in standard Kubernetes, but with some differences due to its lightweight nature.

In K3s, kube-proxy can be configured to run in either "iptables" or "IPVS" mode, depending on the needs of the cluster. In "iptables" mode, kube-proxy creates iptables rules to forward traffic between services, while in "IPVS" mode, it uses the IPVS kernel module to perform the forwarding.

When running in iptables mode, kube-proxy creates a set of iptables rules that allow traffic to be forwarded between services based on their IP addresses and ports. These rules are updated dynamically as services are added, removed, or modified within the cluster.

In IPVS mode, kube-proxy uses the IPVS kernel module to perform load balancing and traffic forwarding between services. IPVS provides more advanced load balancing features than iptables, such as support for different load balancing algorithms and session persistence.

Regardless of the mode it is running in, kube-proxy in K3s also performs other tasks such as updating the cluster's service endpoints and monitoring for changes to the cluster's network topology.

Overall, kube-proxy in K3s plays a critical role in managing network traffic within the cluster, and its efficient operation is essential to ensure the proper functioning of the applications and services running in the cluster.

**Reply by mysticaltech** [reply]
FYI, you can also use Cilium (see `cni_plugin`) with our project, and that by default replaces kube-proxy, and allows for much deeper customization of policies.
